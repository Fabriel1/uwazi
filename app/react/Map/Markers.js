import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Immutable from 'immutable';

function getMarker(entity, templates) {
  const template = templates.find(_t => _t.get('_id') === entity.get('template'));
  const color = templates.indexOf(template);
  const geolocationProp = template.toJS().properties.find(p => p.type === 'geolocation');
  if (geolocationProp) {
    const _entity = entity.toJS();
    const marker = _entity.metadata[geolocationProp.name];
    return marker ? { properties: { entity: _entity, color }, latitude: marker.lat, longitude: marker.lon } : null;
  }

  return null;
}

function getMarkers(entities, templates) {
  return entities.map(entity => getMarker(entity, templates)).toJS().filter(m => m);
}

export const MarkersComponent = ({ children, entities, templates }) => children(getMarkers(entities, templates));

MarkersComponent.defaultProps = {
  entities: Immutable.List(),
};

MarkersComponent.propTypes = {
  children: PropTypes.func.isRequired,
  templates: PropTypes.instanceOf(Immutable.List).isRequired,
  entities: PropTypes.instanceOf(Immutable.List),
};

export const mapStateToProps = ({ templates }) => ({ templates });

export default connect(mapStateToProps)(MarkersComponent);
