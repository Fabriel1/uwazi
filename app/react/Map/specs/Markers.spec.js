import React from 'react';

import { shallow } from 'enzyme';
import Immutable from 'immutable';

import { mapStateToProps, MarkersComponent } from '../Markers.js';

describe('Markers component', () => {
  const state = {
    templates: Immutable.fromJS([{
      _id: 't1',
      properties: [
        { type: 'geolocation', name: 'geoProperty' },
      ]
    }, {
      _id: 't2',
      properties: [
        { type: 'geolocation', name: 'anotherGeoProperty' },
      ]
    }, {
      _id: 't3',
      properties: [
        { type: 'notGeolocation', name: 'notGeo' },
      ]
    }]),
  };

  it('should return parsed markers from entities array', () => {
    const entities = Immutable.fromJS([
      { template: 't1', metadata: { geoProperty: { lat: 7, lon: 13 } } },
      { template: 't1', metadata: { geoProperty: { lat: 5, lon: 22 } } },
      { template: 't3', metadata: { notGeo: { lat: 1977, lon: 7 } } },
      { template: 't2', metadata: { anotherGeoProperty: { lat: 2018, lon: 6 } } },
    ]);

    const props = mapStateToProps(state);
    const resultMarkers = [];

    shallow(
      <MarkersComponent {...props} entities={entities}>
        {markers => markers.map(m => resultMarkers.push(m))}
      </MarkersComponent>
    );

    expect(resultMarkers).toMatchSnapshot();
  });

  it('should return an empty array if no data', () => {
    const props = mapStateToProps(state);
    const resultMarkers = [];

    shallow(
      <MarkersComponent {...props}>
        {markers => markers.map(m => resultMarkers.push(m))}
      </MarkersComponent>
    );

    expect(resultMarkers).toEqual([]);
  });
});
