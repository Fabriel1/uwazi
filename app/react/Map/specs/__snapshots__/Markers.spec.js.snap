// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markers component should return parsed markers from entities array 1`] = `
Array [
  Object {
    "latitude": 7,
    "longitude": 13,
    "properties": Object {
      "color": 0,
      "entity": Object {
        "metadata": Object {
          "geoProperty": Object {
            "lat": 7,
            "lon": 13,
          },
        },
        "template": "t1",
      },
    },
  },
  Object {
    "latitude": 5,
    "longitude": 22,
    "properties": Object {
      "color": 0,
      "entity": Object {
        "metadata": Object {
          "geoProperty": Object {
            "lat": 5,
            "lon": 22,
          },
        },
        "template": "t1",
      },
    },
  },
  Object {
    "latitude": 2018,
    "longitude": 6,
    "properties": Object {
      "color": 1,
      "entity": Object {
        "metadata": Object {
          "anotherGeoProperty": Object {
            "lat": 2018,
            "lon": 6,
          },
        },
        "template": "t2",
      },
    },
  },
]
`;
